bundles/org.eclipse.equinox.p2.repository.tools/src/org/eclipse/equinox/p2/internal/repository/tools/CompositeRepositoryApplication.java

```diff
@@ -106,75 +102,57 @@ public void setFailOnExists(boolean value) {
-		mgr.removeRepository(toInit.getRepoLocation());
-
-		// first try and load to see if one already exists at that location.
-		try {
-			IArtifactRepository repository = mgr.loadRepository(toInit.getRepoLocation(), null);
-			if (validRepositoryLocation(repository) && initDestinationRepository(repository, toInit))
-				return repository;
-			throw new ProvisionException(new Status(IStatus.INFO, Activator.ID, NLS.bind(Messages.CompositeRepository_composite_repository_exists, toInit.getRepoLocation())));
-		} catch (ProvisionException e) {
-			// re-throw the exception if we got anything other than "repo not found"
-			if (e.getStatus().getCode() != ProvisionException.REPOSITORY_NOT_FOUND) {
-				if (e.getCause() instanceof MalformedURLException)
-					throw new ProvisionException(NLS.bind(Messages.exception_invalidDestination, toInit.getRepoLocation()), e.getCause());
-				throw e;
-			}
-		}
-
-		IArtifactRepository source = null;
-		try {
-			if (toInit.getFormat() != null)
-				source = mgr.loadRepository(toInit.getFormat(), 0, null);
-		} catch (ProvisionException e) {
-			//Ignore.
+	protected <T> IRepository<T> initializeDestination(RepositoryDescriptor toInit, IRepositoryManager<T> mgr)
+			throws ProvisionException {
+
+		String repositoryType;
+		String defaultName;
+		if (mgr instanceof IArtifactRepositoryManager) {
+			repositoryType = IArtifactRepositoryManager.TYPE_COMPOSITE_REPOSITORY;
+			defaultName = Messages.CompositeRepository_default_artifactRepo_name;
+		} else if (mgr instanceof IMetadataRepositoryManager) {
+			repositoryType = IMetadataRepositoryManager.TYPE_COMPOSITE_REPOSITORY;
+			defaultName = Messages.CompositeRepository_default_metadataRepo_name;
+		} else {
+			throw new AssertionError("Unsupported repository type: " + mgr.getClass()); //$NON-NLS-1$
 		}
-		//This code assumes source has been successfully loaded before this point
-		try {
-			//No existing repository; create a new repository at destinationLocation but with source's attributes.
-			IArtifactRepository repo = mgr.createRepository(toInit.getRepoLocation(), toInit.getName() != null ? toInit.getName() : (source != null ? source.getName() : Messages.CompositeRepository_default_artifactRepo_name), IArtifactRepositoryManager.TYPE_COMPOSITE_REPOSITORY, source != null ? source.getProperties() : null);
-			initRepository(repo, toInit);
-			return repo;
-		} catch (IllegalStateException e) {
-			mgr.removeRepository(toInit.getRepoLocation());
-			throw e;
-		}
-	}
-
-	@Override
-	protected IMetadataRepository initializeDestination(RepositoryDescriptor toInit, IMetadataRepositoryManager mgr) throws ProvisionException {
 		// remove the repo first.
 		mgr.removeRepository(toInit.getRepoLocation());
 
 		// first try and load to see if one already exists at that location.
 		try {
-			IMetadataRepository repository = mgr.loadRepository(toInit.getRepoLocation(), null);
-			if (!validRepositoryLocation(repository) && initDestinationRepository(repository, toInit))
```

Comment:
`validRepositoryLocation()` either returned true or threw an exception.
Therefore `initDestinationRepository()` was never called for a metadata repository.